ot/mpz-ot-core/src/kos/receiver.rs

```diff
@@ -0,0 +1,493 @@
+        }
+
+        let keys = Vec::from_iter(self.state.keys.drain(..count));
+        let choices = Vec::from_iter(self.state.choices.drain(..count));
+
+        if let Some(tape) = &mut self.tape {
+            let ts = Vec::from_iter(self.state.ts.drain(..count));
+
+            let mut hasher = Hasher::default();
+            ciphertexts.iter().for_each(|ct| {
+                hasher.update(&ct.to_bytes());
+            });
+
+            tape.records.insert(
+                self.state.payload_counter,
+                PayloadRecord {
+                    counter: self.state.ot_counter,
+                    choices: Vec::from_lsb0_iter(choices.iter().copied()),
+                    ts,
+                    keys: keys.clone(),
+                    ciphertext_digest: hasher.finalize().into(),
+                },
+            );
+        }
+
+        let plaintexts = keys
+            .into_iter()
+            .zip(choices)
+            .zip(ciphertexts.chunks(2))
+            .map(|((key, c), ct)| if c { key ^ ct[1] } else { key ^ ct[0] })
+            .collect();
+
+        self.state.ot_counter += count;
+        self.state.payload_counter += 1;
+
+        Ok(plaintexts)
+    }
+
+    /// Checks the purported messages against the receiver's protocol tape, using the sender's
+    /// base choices `delta`.
+    ///
+    /// # ⚠️ Warning ⚠️
+    ///
+    /// The authenticity of `delta` must be established outside the context of this function. This
+    /// can be achieved using verifiable OT for the base choices.
+    ///
+    /// # Arguments
+    ///
+    /// * `index` - The index of the payload.
+    /// * `delta` - The sender's base choices.
+    /// * `purported_msgs` - The purported messages sent by the sender.
+    pub fn verify(
+        &self,
+        index: usize,
+        delta: Block,
+        purported_msgs: &[[Block; 2]],
+    ) -> Result<(), ReceiverError> {
+        let Some(tape) = &self.tape else {
+            return Err(ReceiverVerifyError::TapeNotRecorded)?;
+        };
+
+        let Some(record) = tape.records.get(&index) else {
+            return Err(ReceiverVerifyError::InvalidPayloadIndex)?;
+        };
+
+        let PayloadRecord {
+            counter,
+            choices,
+            ts,
+            keys,
+            ciphertext_digest,
+        } = record;
+
+        // Here we compute the complementary key to the one used earlier in the protocol.
+        //
+        // From this, we encrypt the purported messages and check that the ciphertext digests match.
+        let cipher = &(*FIXED_KEY_AES);
+        let mut hasher = Hasher::default();
+        for (j, (((c, t), key), msgs)) in choices
+            .iter_lsb0()
+            .zip(ts)
+            .zip(keys)
+            .zip(purported_msgs)
+            .enumerate()
+        {
+            let j = Block::new(((counter + j) as u128).to_be_bytes());
+            let key_ = cipher.tccr(j, *t ^ delta);
+
+            let (ct0, ct1) = if c {
```

Comment:
This code logic doesn't hold when random choice `c` is 1 ( `*key` is `key1`) but the actual choice is 0. Then the Sender's flip==1 and he will send (msg0^key1, msg1^key0)
but this code expects that the Sender sends (msg0^key0, msg1^key1)

Reply:
@sinui0, I'm still seeing that it is a problem. Maybe I could enumerate the steps and you'll tell me which one is incorrect:

- assume `R`'s random choice bit `c` is 1, then his `key` is `key_1`
- assume `R`'s actual choice bit is 0, then during derandomization `R` sends the flip bit `1`
- `S` performs the flip and sends back `ct_0` = `msg_0 ^ key_1` and `ct_1` = `msg_1 ^ key_0`
- `R` writes adds (`ct_0`, `ct_1`) to the tape

- during the verification `R` computes a complementary key `key_` which is `key_0`
- acc.to the code since his random choice bit `c` was 1, he executes the `if c {}` branch
- he computes `ct0 = msgs[0] ^ key_`, which is `msg_0 ^ key_0` 

`ct0` does not match `ct_0` which `S` sent earlier